packages/opentelemetry-node-tracer/src/instrumentation/PluginLoader.ts

```diff
@@ -0,0 +1,146 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Logger, Plugin, Tracer } from '@opentelemetry/types';
+import * as hook from 'require-in-the-middle';
+import * as utils from './utils';
+
+// States for the Plugin Loader
+export enum HookState {
+  UNINITIALIZED,
+  ENABLED,
+  DISABLED,
+}
+
+interface PluginNames {
+  [pluginName: string]: string;
+}
+
+// tslint:disable-next-line:no-any
+export type PluginType = Array<Plugin<any>>;
+
+/**
+ * The PluginLoader class can load instrumentation plugins that
+ * use a patch mechanism to enable automatic tracing for
+ * specific target modules.
+ */
+export class PluginLoader {
+  /** A list of loaded plugins. */
+  private _plugins: PluginType = [];
+  /**
+   * A field that tracks whether the require-in-the-middle hook has been loaded
+   * for the first time, as well as whether the hook body is activated or not.
+   */
+  private _hookState = HookState.UNINITIALIZED;
+
+  /** Constructs a new PluginLoader instance. */
+  constructor(readonly tracer: Tracer, readonly logger: Logger) {}
+
+  get state(): HookState {
+    return this._hookState;
+  }
+
+  get plugins(): PluginType {
+    return this._plugins;
+  }
+
+  /**
+   * Loads a list of plugins. Each plugin module should implement the core
+   * {@link Plugin} interface and export an instance named as 'plugin'. This
+   * function will attach a hook to be called the first time the module is
+   * loaded.
+   * @param modulesToPatch A list of plugins.
+   */
+  load(modulesToPatch: string[]): PluginLoader {
+    if (this._hookState === HookState.UNINITIALIZED) {
+      const plugins = modulesToPatch.reduce(
+        (plugins: PluginNames, moduleName: string) => {
+          plugins[moduleName] = utils.defaultPackageName(moduleName);
+          return plugins;
+        },
+        {} as PluginNames
+      );
+      const modulesToHook = Object.keys(plugins);
+      // Do not hook require when no module is provided. In this case it is
+      // not necessary. With skipping this step we lower our footprint in
+      // customer applications and require-in-the-middle won't show up in CPU
+      // frames.
+      if (modulesToHook.length === 0) {
+        this._hookState = HookState.DISABLED;
+        return this;
+      }
+
+      // Enable the require hook.
+      hook(modulesToHook, (exports, name, baseDir) => {
+        if (this._hookState !== HookState.ENABLED) return exports;
+
+        const moduleName = plugins[name];
+        // Get the module version.
+        const version = utils.getPackageVersion(this.logger, baseDir as string);
+        this.logger.info(
+          `PluginLoader#load: trying loading ${name}.${version}`
+        );
+        if (!version) return exports;
```

Comment:
I would check if version and supportedVersions are satisfied (with semver)

Reply:
Sure, Added `@todo` for now.